x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -410,6 +425,32 @@ private Map<String, Object> buildUserMetadata(JWTClaimsSet claimsSet) {
         return Map.copyOf(metadata);
     }
 
+    private String buildTokenPrincipal(JWTClaimsSet jwtClaimsSet) {
+        // If Issuer is not found, throw an error because it does not have fallback and no other realm can work with it.
+        final String issuer = jwtClaimsSet.getIssuer();
+        if (Strings.hasText(issuer) == false) {
+            throw new IllegalArgumentException("Issuer claim 'iss' is missing.");
+        }
+
+        // If either Subject or Audience is not found, yield and let other realms try it again since they may have different fallbacks
+        final Map<String, String> fallbackClaimNames = jwtAuthenticator.getFallbackClaimNames();
+        final FallbackableClaim subClaim = new FallbackableClaim("sub", fallbackClaimNames, jwtClaimsSet);
+        final String subject = subClaim.getStringClaimValue();
+        if (false == Strings.hasText(subject)) {
+            logger.debug("claim [{}] is missing for building token principal for realm [{}]", subClaim, name());
+            return null;
+        }
+
+        final FallbackableClaim audClaim = new FallbackableClaim("aud", fallbackClaimNames, jwtClaimsSet);
+        final List<String> audiences = audClaim.getStringListClaimValue();
+        if (audiences == null || audiences.isEmpty()) {
+            logger.debug("claim [{}] is missing for building token principal for realm [{}]", audClaim, name());
+            return null;
+        }
+
+        return issuer + "/" + String.join(",", new TreeSet<>(audiences)) + "/" + subject;
+    }
```

Comment:
I have some concerns about this. I need some time to think about it, and come back.